File: r.go
package c2rust

import (
	"fmt"
	"strings"
)

type RVar struct {
	label   string
	vartype string
	mutable bool
}

type RVarDecl struct {
	rvar        RVar
	initializer string
}

type RAssignment struct {
	rvar  RVar
	right string
}

type RFunction struct {
	label      string
	args       []RVar
	body       []RWriter
	returntype string
}

type RWriter interface {
	write()
}

func (f *RFunction) write() {
	argstr := []string{}
	for _, x := range f.args {
		if x.mutable {
			argstr = append(argstr, fmt.Sprintf("mut %s: %s", x.label, x.vartype))
		} else {
			argstr = append(argstr, fmt.Sprintf("%s: %s", x.label, x.vartype))
		}
	}
	if f.returntype != "" {
		fmt.Printf("fn %s(%s) -> %s {\n", f.label, strings.Join(argstr, ", "), f.returntype)
	} else {
		fmt.Printf("fn %s(%s) {\n", f.label, strings.Join(argstr, ", "))
	}
	for _, x := range f.body {
		x.write()
	}
	fmt.Println("}")
}

func (d *RVarDecl) write() {
	if d.rvar.mutable {
		fmt.Printf("let mut %s: %s", d.rvar.label, d.rvar.vartype)
	} else {
		fmt.Printf("let %s: %s", d.rvar.label, d.rvar.vartype)
	}
	if d.initializer != "" {
		fmt.Printf(" = %s", d.initializer)
	}
	fmt.Println()
}

File: r.rs
// r.go

pub struct RVar {
	label:   String,
	vartype: String,
	mutable: bool
}

pub struct RVarDecl {
	rvar:        RVar,
	initializer: String
}

pub struct RAssignment {
	rvar:  RVar,
	right: String
}

pub struct RFunction {
	label:      String,
	args:       Vec<RVar>,
	body:       Vec<RWriter>,
	returntype: String
}

trait RWriter {
	fn write(&self);
}

impl RWriter for RFunction {

    fn write(f: *const RFunction) {
    	let mut argstr = Vec::new();
    	for x in 0..f.args.len(){
    		if x.mutable {
    			argstr.push(format!("mut {}: {}", x.label, x.vartype));
    		} else {
    			argstr.push(format!("{}: {}", x.label, x.vartype));
    		}
    	}
    	if f.returntype != "" {
    		print!("fn {}({}) -> {} {{\n", f.label, format!("{}{}",argstr, ", "), f.returntype)
    	} else {
    		print!("fn {}({}) {{\n", f.label, format!("{}{}",argstr, ", "))
    	}
    	for x in 0..f.body.len() {
    		x.write()
    	}
    	println!("}}");
    }
}

impl RWriter for RVarDecl {
    fn write(d: *const RVarDecl) {
    	if d.rvar.mutable {
    		print!("let mut {}: {}", d.rvar.label, d.rvar.vartype);
    	} else {
    		print!("let {}: {}", d.rvar.label, d.rvar.vartype);
    	}
    	if d.initializer != "" {
    		print!(" = {}", d.initializer);
    	}
    	println!("");
    }
}
